os/fs/fat/fs_fat32.c

```diff
@@ -0,0 +1,2678 @@
+		 */
+
+		goto errout_with_semaphore;
+	}
+
+	/* Create an instance of the file private date to describe the opened
+	 * file.
+	 */
+
+	ff = (FAR struct fat_file_s *)kmm_zalloc(sizeof(struct fat_file_s));
+	if (!ff) {
+		ret = -ENOMEM;
+		goto errout_with_semaphore;
+	}
+
+	/* Create a file buffer to support partial sector accesses */
+
+	ff->ff_buffer = (FAR uint8_t *)fat_io_alloc(fs->fs_hwsectorsize);
+	if (!ff->ff_buffer) {
+		ret = -ENOMEM;
+		goto errout_with_struct;
+	}
+
+	/* Initialize the file private data (only need to initialize non-zero elements) */
+
+	ff->ff_oflags = oflags;
+
+	/* Save information that can be used later to recover the directory entry */
+
+	ff->ff_dirsector = fs->fs_currentsector;
+	ff->ff_dirindex = dirinfo.dir.fd_index;
+
+	/* File cluster/size info */
+
+	ff->ff_startcluster = ((uint32_t)DIR_GETFSTCLUSTHI(direntry) << 16) | DIR_GETFSTCLUSTLO(direntry);
+
+	ff->ff_currentcluster = ff->ff_startcluster;
+	ff->ff_sectorsincluster = fs->fs_fatsecperclus;
+	ff->ff_size = DIR_GETFILESIZE(direntry);
+
+	/* Attach the private date to the struct file instance */
+
+	filep->f_priv = ff;
+
+	/* Then insert the new instance into the mountpoint structure.
+	 * It needs to be there (1) to handle error conditions that effect
+	 * all files, and (2) to inform the umount logic that we are busy
+	 * (but a simple reference count could have done that).
+	 */
+
+	ff->ff_next = fs->fs_head;
+	fs->fs_head = ff;
+
+	fat_semgive(fs);
+
+	/* In write/append mode, we need to set the file pointer to the end of the file */
+
+	if ((oflags & (O_APPEND | O_WRONLY)) == (O_APPEND | O_WRONLY)) {
+		off_t offset = fat_seek(filep, ff->ff_size, SEEK_SET);
+		if (offset < 0) {
+			kmm_free(ff);
+			return (int)offset;
+		}
+	}
+
+	return OK;
+
+	/* Error exits -- goto's are nasty things, but they sure can make error
+	 * handling a lot simpler.
+	 */
+
+errout_with_struct:
+	kmm_free(ff);
+
+errout_with_semaphore:
+	fat_semgive(fs);
+	return ret;
+}
+
+/****************************************************************************
+ * Name: fat_close
+ ****************************************************************************/
+
+static int fat_close(FAR struct file *filep)
+{
+	FAR struct inode *inode;
+	FAR struct fat_file_s *ff;
+	FAR struct fat_file_s *currff;
+	FAR struct fat_file_s *prevff;
+	FAR struct fat_mountpt_s *fs;
+	int ret = OK;
+
+	/* Sanity checks */
+
+	DEBUGASSERT(filep->f_priv != NULL && filep->f_inode != NULL);
+
+	/* Recover our private data from the struct file instance */
+
+	ff = filep->f_priv;
+	inode = filep->f_inode;
+	fs = inode->i_private;
+
+	DEBUGASSERT(fs != NULL);
+
+	/* Check for the forced mount condition */
+
+	if ((ff->ff_bflags & UMOUNT_FORCED) == 0) {
+		/* Do not check if the mount is healthy.  We must support closing of
+		 * the file even when there is healthy mount.
+		 */
+
+		/* Synchronize the file buffers and disk content; update times */
+
+		ret = fat_sync(filep);
```

Comment:
No need to check whether sync is successful?

Reply:
It's better to have it.